File: src/infrastructure/tools/file/file-info.ts
/**
 * file_info — 获取文件/目录元信息（类型、大小、访问/修改时间）
 */
import * as tauriFs from '@tauri-apps/plugin-fs'
import { t } from '@/ui/i18n'
import { toolRegistry } from '@/domain/tools'
import type { ToolContext, ToolExecutor, ToolResult } from '@/domain/tools/types'
import { securityService } from '@/services/security-service'
import { formatSize, isTauriFsAvailable } from './common'

/** 本地时间格式 %Y-%m-%d %H:%M:%S（与 Rust `format_system_time` 一致） */
function formatSystemTime(d: Date): string {
  const p = (n: number) => String(n).padStart(2, '0')
  return `${d.getFullYear()}-${p(d.getMonth() + 1)}-${p(d.getDate())} ${p(d.getHours())}:${p(d.getMinutes())}:${p(d.getSeconds())}`
}

toolRegistry.register(
    'file_info',
    (async (
      args: Record<string, any>,
      ctx: ToolContext,
    ): Promise<ToolResult | string> => {
    if (!isTauriFsAvailable())
      return '[file_info] Error: not running in a Tauri environment'

    const fullPath = await securityService.resolveSafePath(
      args.path as string,
      'r',
      ctx.sessionId,
    )

    try {
      const exists = await tauriFs.exists(fullPath)
      if (!exists) return `Error: path does not exist — ${fullPath}`

      const stat = await tauriFs.stat(fullPath)
      const atimeMs = stat.atime ? stat.atime.getTime() : null
      const mtimeMs = stat.mtime ? stat.mtime.getTime() : null

      // ⚠️ 模型侧固定英文（P4b/D2-A）；UI 侧走 uiData 由组件按界面语言渲染。
      // 文案与 Rust `native_tools/file/file_info.rs` 逐字一致（铁律 1）。
      const content = [
        `📋 ${fullPath}`,
        `  Type: ${stat.isDirectory ? '📁 Directory' : '📄 File'}`,
        stat.size !== undefined ? `  Size: ${formatSize(stat.size)}` : '',
        atimeMs != null ? `  Accessed: ${formatSystemTime(new Date(atimeMs))}` : '',
        mtimeMs != null ? `  Modified: ${formatSystemTime(new Date(mtimeMs))}` : '',
      ]
        .filter(Boolean)
        .join('\n')

      return {
        content,
        uiData: {
          path: fullPath,
          isDirectory: !!stat.isDirectory,
          sizeBytes: stat.size ?? null,
          atimeMs,
          mtimeMs,
        },
      }
    } catch (e: any) {
      return `Error: failed to get file info — ${e.message || String(e)}`
    }
  }) as ToolExecutor,
    t('文件信息'),
)

File: src-tauri/virlen-core/src/agent/native_tools/file/file_info.rs
//! `file_info` 工具（原生）— 返回文件/目录的存在性、类型、大小与访问/修改时间。

use crate::agent::native_tools::common::{arg_str, resolve_safe_path};
use crate::agent::native_tools::{NativeToolCtx, NativeToolOutcome};
use serde_json::{json, Value};

use super::common::{format_size, format_system_time};

pub(crate) async fn file_info_tool(
    ctx: &NativeToolCtx<'_>,
    args: &Value,
) -> Result<NativeToolOutcome, String> {
    let path = arg_str(args, "path").unwrap_or_default();
    let full_path = resolve_safe_path(&path, "r", ctx.security)?;

    if !std::path::Path::new(&full_path).exists() {
        return Ok(NativeToolOutcome::Value {
            content: format!("Error: path does not exist — {}", full_path),
            ui_data: None,
        });
    }

    let metadata = std::fs::metadata(&full_path)
        .map_err(|e| format!("Error: failed to get file info — {}", e))?;

    let is_dir = metadata.is_dir();
    let size = metadata.len();
    let atime = metadata
        .accessed()
        .ok()
        .map(|t| format_system_time(t))
        .unwrap_or_default();
    let mtime = metadata
        .modified()
        .ok()
        .map(|t| format_system_time(t))
        .unwrap_or_default();

    let lines = vec![
        format!("📋 {}", full_path),
        format!("  Type: {}", if is_dir { "📁 Directory" } else { "📄 File" }),
        format!("  Size: {}", format_size(size as usize)),
        if atime.is_empty() {
            String::new()
        } else {
            format!("  Accessed: {}", atime)
        },
        if mtime.is_empty() {
            String::new()
        } else {
            format!("  Modified: {}", mtime)
        },
    ];

    // UI 侧结构化元信息（时间戳毫秒，由组件按界面语言本地化）
    let atime_ms = metadata
        .accessed()
        .ok()
        .and_then(|t| t.duration_since(std::time::UNIX_EPOCH).ok())
        .map(|d| d.as_millis() as u64);
    let mtime_ms = metadata
        .modified()
        .ok()
        .and_then(|t| t.duration_since(std::time::UNIX_EPOCH).ok())
        .map(|d| d.as_millis() as u64);

    Ok(NativeToolOutcome::Value {
        content: lines.into_iter().filter(|l| !l.is_empty()).collect::<Vec<_>>().join("\n"),
        ui_data: Some(json!({
            "path": full_path,
            "isDirectory": is_dir,
            "sizeBytes": size,
            "atimeMs": atime_ms,
            "mtimeMs": mtime_ms,
        })),
    })
}
